### PR TITLE
gh-107557: Remove unnecessary SAVE_IP instructions

### PR DESCRIPTION
**Pull request by gvanrossum**
As a final optimization pass, squeeze out any SAVE_IP instructions that are not needed.

For example, if we have
```
SAVE_IP
LOAD_FAST
SAVE_IP
LOAD_GLOBAL
```
we can remove the first `SAVE_IP`, since `LOAD_FAST` can neither deoptimize nor error out; but we cannot remove the second `SAVE_IP`, since `LOAD_GLOBAL` can error out.

This pass requires new per-opcode flags, `HAS_ERROR_FLAG` and `HAS_DEOPT_FLAG`, which must be defined for regular ops as well as for uops. As a bonus we also remove redundant `NOP` opcodes.

(It may seem suboptimal to call `squeeze_stubs()` twice -- once after constructing an initial trace if it has stubs, once after removing unneeded uops; but to avoid this we'd need to change a bunch of APIs to return where the gap is. This seems simpler.)

* Issue: gh-107557

### CODE REVIEW

File: Python/optimizer.c [hunk 1]
```diff
@@ -779,6 +800,76 @@ translate_bytecode_to_trace(
 #undef DPRINTF
 }
 
+static int
+remove_unneeded_uops(_PyUOpInstruction *trace, int trace_length)
+{
+    // Stage 1: Replace unneeded SAVE_IP uops with NOP.
+    // Note that we don't enter stubs, those SAVE_IPs are needed.
+    int last_save_ip = -1;
```
Comment (Fidget-Spinner) [inline Python/optimizer.c:808]: I think it might be clearer if we call this variable `has_previous_save_ip` and make it a boolean.

**Review by Fidget-Spinner — APPROVED** [review]

File: Python/optimizer.c [hunk 2]
```diff
@@ -779,6 +800,76 @@ translate_bytecode_to_trace(
...
+            }
+            need_ip = false;
+            last_save_ip = pc;
+        }
+        else if (opcode == JUMP_TO_TOP || opcode == EXIT_TRACE) {
+            last_instr = pc + 1;
+            break;
+        }
+        else {
+            // If opcode metadata flags has ERROR or DEOPT, set need_up to true
```
Comment (iritkatriel) [inline Python/optimizer.c]:
```suggestion
            // If opcode has ERROR or DEOPT, set need_ip to true
```

File: Python/optimizer.c [hunk 3]
```diff
@@ -372,6 +372,32 @@ static PyTypeObject UOpExecutor_Type = {
     .tp_as_sequence = &uop_as_sequence,
 };
 
+static int
+squeeze_stubs(
```
Comment (iritkatriel) [inline Python/optimizer.c]:
```suggestion
shift_stubs(
```

(the name implied to me that something happens to the size of the stubs).

File: Python/optimizer.c (hunk 1 above)
Comment (iritkatriel) [reply Python/optimizer.c:808]: That would imply it's boolean, but it's an index.

File: Python/optimizer.c [hunk 4]
```diff
@@ -779,6 +800,76 @@ translate_bytecode_to_trace(
...
+            last_instr = pc + 1;
+            break;
+        }
+        else {
+            // If opcode metadata flags has ERROR or DEOPT, set need_up to true
+            if (_PyOpcode_opcode_metadata[opcode].flags & (HAS_ERROR_FLAG | HAS_DEOPT_FLAG)) {
+                need_ip = true;
+            }
+        }
+    }
```
Comment (iritkatriel) [inline Python/optimizer.c:834]: assert here that ``last_instr >= 0``?

File: Tools/cases_generator/flags.py [hunk 5]
```diff
@@ -40,11 +42,19 @@ def fromInstruction(instr: parsing.Node) -> "InstructionFlags":
             )
             and not has_free,
             HAS_EVAL_BREAK_FLAG=variable_used(instr, "CHECK_EVAL_BREAKER"),
+            HAS_DEOPT_FLAG=variable_used(instr, "DEOPT_IF"),
+            HAS_ERROR_FLAG=(
+                variable_used(instr, "ERROR_IF")
+                or variable_used(instr, "error")
+                or variable_used(instr, "pop_1_error")
+                or variable_used(instr, "exception_unwind")
+                or variable_used(instr, "resume_with_error")
```
Comment (iritkatriel) [inline Tools/cases_generator/flags.py:51]: We could make ``variable_used`` accept a sequence of strings to check.

File: Python/optimizer.c (hunk 3 above)
Comment (gvanrossum) [reply Python/optimizer.c]: How about `move_stubs()` (echoing that it uses `memmove()`).

File: Python/optimizer.c (hunk 4 above)
Comment (gvanrossum) [reply Python/optimizer.c:834]: No, I can just initialize it to 0 though. If the trace is empty we shouldn't get here but the code here won't do anything.

File: Tools/cases_generator/flags.py (hunk 5 above)
Comment (gvanrossum) [reply Tools/cases_generator/flags.py:51]: We could, but why bother? Plus, I'd kind of like `variable_used(instr, X, Y, Z)` to look for the *sequence* of identifiers `X Y Z`. (And I'd also like to allow keywords, e.g. `goto`.)

**Review by gvanrossum — COMMENTED** [review]
Thanks for the reviews! I've changed the function name, shortened the comment, and changed the initialization of last_instr. I'll merge once tests pass.

**Comment by gvanrossum** [comment]
[Benchmarking results](https://github.com/faster-cpython/benchmarking-public/tree/main/results/bm-20230828-3.13.0a0-1ff7ef4) show this would be 4% slower than main if the optimizer was always on (it isn't).

File: Python/optimizer.c (hunk 1 above)
Comment (Fidget-Spinner) [reply Python/optimizer.c:808]: :facepalm: I thought I only saw it being used as a boolean, but then saw it's used to index the trace. Yeah sorry my mistake!

**Pull request merged**